card/src/main/java/com/adyen/checkout/card/CardView.kt

```diff
@@ -163,27 +166,30 @@ class CardView @JvmOverloads constructor(context: Context, attrs: AttributeSet?
             if (!it.cardNumberState.validation.isValid()) {
                 isErrorFocused = true
                 binding.editTextCardNumber.requestFocus()
-                setCardNumberError(R.string.checkout_card_number_not_valid)
+                setCardNumberError((it.cardNumberState.validation as Validation.Invalid).reason)
             }
             if (!it.expiryDateState.validation.isValid()) {
                 if (!isErrorFocused) {
                     isErrorFocused = true
                     binding.textInputLayoutExpiryDate.requestFocus()
                 }
-                binding.textInputLayoutExpiryDate.error = mLocalizedContext.getString(R.string.checkout_expiry_date_not_valid)
+                val errorReasonResId = (it.expiryDateState.validation as Validation.Invalid).reason
```

Comment:
I feel like this could be improved, the if condition checks `isValid()` which internally checks `is Validation.Valid`, but we have to cast again to `Validation.Invalid` here which doesn't look great IMO.

One fix could be something like this:

```kotlin
val expiryDateValidation = it.expiryDateState.validation
if (expiryDateValidation is Validation.Invalid) {
                if (!isErrorFocused) {
                    isErrorFocused = true
                    binding.textInputLayoutExpiryDate.requestFocus()
                }
                val errorReasonResId = expiryDateValidation.reason // no need to cast anymore
```

On the java side we cannot benefit from the automatic type inference like kotlin, but we can still maybe extract the variable at least:

```java
val cardNumberValidation = outputData.getCardNumberField().getValidation();
if (!cardNumberValidation.isValid()) {
            isErrorFocused = true;
            mCardNumberEditText.requestFocus();
            final int errorReasonResId = ((Validation.Invalid) cardNumberValidation).getReason();
            setCardNumberError(errorReasonResId);
        }
```

The `isValid()` method would still be useful in other places but not if we need to get the reason. What do you think?